VNFS Update to support SLES15 and cross-distro vnfs

jcsiadal opened the pull request:
Quick background: I'm doing work with OpenSUSE 15 and SLES 15. OpenSUSE was working, but SLES has no online repos and requires use of ISOs. Those ISOs include 30+ repos, but I still wanted to use the INSTALL_ISO capability. In addition, I wanted a SLES head node with OpenSUSE compute nodes.

Posted here as much to start discussion as to implement.

Here's a summary of changes:
1) wwmkvnfs script updated to make better use of verbose and debug options. Getting SLES to work required a lot of debug and it's useful having the ability to insert permanent output points in the scripts. Also added a switch to disable GPG check on builds, without turning it off in the repos.

2) The scripts are more quiet by default.

3) Both Red Hat and SUSE builds are now fully independent of the head node. On the SLES head, I was able to build working OpenSUSE 42 and 15, and CentOS 8 computes. I couldn't find a Zypper rpm for CentOS, which would be needed to go the other way. Still need more testing with Red Hat, but I tested CentOS7 computes on a RHEL8 head. I would like to update Ubuntu/Debian to support cross-VNFS builds with SUSE and Red Hat.

4) Fixed the INSTALL_ISO mounting and repo scans.

5) Added a variable to disable gpg checking on compute node repos.

6) Added a variable to pin a single kernel version on SUSE (can be overridden) I believe I can do something like this on CentOS and Ubuntu too.

7) Turned off weak dependencies on compute nodes. In addition to bloating the compute images, this has caused problems with some packages on SLES. For example, extra driver packages "Suggest" kernel-default, which causes it to install.

8) Deleted a few old templates. I don't think anyone would get around to testing them.

9) Added SLES 15 template

10) Minor fixups here and there.

jcsiadal committed:
VNFS update to support SLES15

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jmstover commented on vnfs/bin/wwmkchroot.in:
```diff
@@ -38,23 +40,34 @@ usage() {
             usage
             exit 0
         ;;
-        a)
```
This can probably be removed. I don't think it was used much. IIRC, we had this as a way for simple passing of a value into the `*.tmpl` file...  So, `$var` was used in the template, and you'd do:

```
wwmkchroot -a bob [...]
```

To pass the `bob` value in.  Hackish...

jcsiadal replied:
I've seen it used. Close this comment?

jcsiadal replied:
I meant "never seen -a used".

bensallen commented on vnfs/libexec/wwmkchroot/functions:
```diff
@@ -1,6 +1,12 @@
-eval `wwconfig -as`
+#!/bin/bash
```
Unless this PR intends to execute the functions scripts directly, there's no need to add a #! line back into a script that is only ever sourced.

jcsiadal replied:
Up to you guys if I need to pull this. It's definitely handy with shell checker and any autoformatting editor will handle the file correctly.

jmstover replied:
@bensallen ... I'm not partial either way. For "correctness" it probably isn't needed. But in all honestly, it doesn't hurt anything to have it. Would be nice to hear @jjolly 's thoughts.

bensallen commented on vnfs/libexec/wwmkchroot/include-rhel:
```diff
@@ -1,30 +1,56 @@
-    YUM_CMD="yum -y -c $CHROOTDIR/$YUM_CONF --tolerant --installroot $CHROOTDIR $EXTRA_OPTS"
-    YUM_REPOCMD="yum-config-manager -y -c $CHROOTDIR/$YUM_CONF --tolerant --installroot $CHROOTDIR $EXTRA_OPTS"
-fi
+#!/bin/bash
```
Same as above, no need to add a #! line back into a script that is only ever sourced.

bensallen commented on vnfs/libexec/wwmkchroot/include-suse:
```diff
@@ -1,10 +1,26 @@
-ZYP_CMD="zypper --root $CHROOTDIR -n --gpg-auto-import-keys"
+#! /bin/bash
```
No need to add a #! line back into a script that is only ever sourced.

jcsiadal commented:
Thanks for looking this over. I agree that the shebang isn't needed during runtime. I added it because ShellCheck complained the format was ambiguous. More importantly for me, vim in OpenSUSE wouldn't autoformat shell scripts until I added it.

jmstover commented:
> More importantly for me, vim in OpenSUSE wouldn't autoformat shell scripts until I added it.

Weird... would adding a modeline entry in work instead? Like what's in `vnfs/libexec/wwmkchroot/include-deb`?

```
# vim: filetype=sh:syntax=sh:expandtab:ts=4:sw=4:
```

jcsiadal commented:
Once I reconfigure my .vimrc, the modeline does work. I can also force shellcheck to a specific dialect of shell. On the other hand, a shebang doesn't impact anything and it's helpful. My preference is to include both.

jcsiadal committed:
Ubuntu support added

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

bensallen commented:
@jjolly In #151 you suggested we remove the shebang lines from the scripts that are only ever sourced. I'd like to get your input on @jcsiadal's request to add them back so editors better detect that the file is a bash script.

jmstover commented:
In general, I think it should be fine without (except the detection). I don't *think* we're using any bashisms anywhere... So, really they *should* be /bin/sh scripts ...  If they were attempted to be ran without the shebang `/bin/sh` should be the shell used.

But... they also won't run under dash ... For example, we have:
```
. include-rhel
```

Dash is going to search the `PATH` for `include-rhel` ... not try sourceing the file in the current directory (Is that POSIX?).

So, to do what we're wanting ... we'd need to explicitly put a path in there like:
```
. ./include-rhel
```
I used `dash-0.5.10.2-1.el7.x86_64` on SL 7.4 when testing this.

jcsiadal committed:
Merge branch 'development' into slesvnfs

jcsiadal commented:
Hey guys, I'd like to revisit this PR. I've merged new changes to the dev branch.

Regarding the shebang in includes, just let me know whether to pull them or if we can keep. They are useful for outside editors and checkers, but they're certainly not required.

However, I do use arrays. They're much more reliable when manipulating data containing quotes and for space-separated lists. And since POSIX doesn't support arrays, sh shouldn't allow them. I'm also a fan of using bash string expansions.

jmstover commented on vnfs/libexec/wwmkchroot/functions:
```diff
@@ -245,13 +273,20 @@ configure_services() {
+        if [ -n "$DEBUG" ]; then
+            echo "WARNING: Debug enabled; leaving media mounts in place!"
+        else
+            rm -f $MEDIA_TMP/*
```
You're using `$MEDIA_TMP/*` here ... Above you're setting:
```
MEDIA_MOUNTPATH="${CHROOTDIR}${MEDIA_TMP}"
```
Which is to be used?

jcsiadal replied:
It's actually both. I added some comments to the variable declaration that will hopefully clarify this. Basically, everything is mounted under the chroot, but the host need to access the media to bootstrap the image. The same directory is created on the host with links to the media in the chroot.

In case you're wonding why I make the symlinks to the files: I could have just created one symlink to the media directory, but I would need remove any directory with that name first which seemed too destructive for me.

jmstover commented:
> However, I do use arrays.

That should be fine here .... `wwmkchroot` itself has a bash shebang line, so all of these _should_ be ran under bash in normal usage. We just try to keep the bashism's down as much as possible. Never know when you may need to change the shell it's running under.

jcsiadal committed:
Add comments to vars

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

bensallen reviewed (approved):
I think this has sat long enough. Let's merge it. If someone finds an issue with the bashisms or #! lines we can open a new PR to fix.
